# Problem Definition

We need to solve the following multi-objective optimization problem using the concession method without using ready-made libraries or packages in Python.

# Objective Functions:

\begin{align*}
f_1 &= x_1 + 4x_2 \quad \text{to maximize} \\
f_2 &= bx_1 + \frac{b}{2} x_2 \quad \text{to maximize}
\end{align*}

# Constraints:
\begin{align*}
x_1 + 4x_2 &\leq 24 \\
x_1 + x_2 &\leq 10 \\
x_1 - 5x_2 &\leq 5 \\
x_1, x_2 &\geq 0
\end{align*}



Given that:


*   𝑏 is the option number, let's assume the option number is 1, hence 𝑏 = 1.
*   𝑑 = $\frac{b}{2}$ = 0.5.





# Concession Method:

If the concession is not determined, we take it as 7%.

# Steps:

1.   Define the Objective Functions and Constraints:

*   Implement the objective functions 𝑓1 and 𝑓2.
*   Implement the constraints as a function that checks if they are satisfied.

2.   Penalty Function:

*  Create a penalty function to impose infinite penalties for violating constraints.

3.   Concession Method:

*   Iterate over possible values for 𝑥1 and 𝑥2 within a feasible range.
*   Evaluate the objective functions and apply the constraints.
*   Apply the concession to ensure a trade-off between 𝑓1 and 𝑓2.


In [ ]:
import numpy as np

# Define parameters
b = 1
d = b / 2
concession = 0.07

# Define objective functions
def f1(x1, x2):
    return x1 + 4 * x2

def f2(x1, x2):
    return b * x1 + d * x2

# Define constraints
def constraints(x1, x2):
    return [
        x1 + 4 * x2 <= 24,
        x1 + x1 <= 10,
        x1 - 5 * x2 <= 5,
        x1 >= 0,
        x2 >= 0
    ]

# Create a penalty function for the constraints
def penalty(x1, x2):
    penalties = [0 if constraint else np.inf for constraint in constraints(x1, x2)]
    return sum(penalties)

# Create the concession method function
def concession_method():
    best_f1 = -np.inf
    best_f2 = -np.inf
    best_x1 = 0
    best_x2 = 0

    for x1 in np.linspace(0, 24, 500):
        for x2 in np.linspace(0, 24, 500):
            if penalty(x1, x2) == 0:
                current_f1 = f1(x1, x2)
                current_f2 = f2(x1, x2)

                if current_f1 > best_f1 and current_f2 > best_f2 * (1 - concession):
                    best_f1 = current_f1
                    best_f2 = current_f2
                    best_x1 = x1
                    best_x2 = x2

    return best_x1, best_x2, best_f1, best_f2

In [ ]:
# Execute the concession method
best_x1, best_x2, best_f1, best_f2 = concession_method()

print(f"Best values found: x1 = {best_x1}, x2 = {best_x2}")
print(f"Best f1: {best_f1}, Best f2: {best_f2}")

Best values found: x1 = 0.14428857715430862, x2 = 5.963927855711423
Best f1: 24.0, Best f2: 3.12625250501002


# Verification of Results
Given the results obtained from the Python code:

*   best_x1 = 0.14428857715430862
*   best_x2 = 5.963927855711423


We need to verify if these values satisfy all the constraints and if they maximize the objective functions.

# Constraints Verification

\begin{align*}
1. & \quad x_1 + 4x_2 \leq 24 \\
   & \quad 0.14428857715430862 + 4 \times 5.963927855711423 = 24 \leq 24 \quad \text{(Satisfies)} \\
2. & \quad x_1 + x_2 \leq 10 \\
   & \quad 0.14428857715430862 + 5.963927855711423 = 6.108216432865732 \leq 10 \quad \text{(Satisfies)} \\
3. & \quad x_1 - 5x_2 \leq 5 \\
   & \quad 0.14428857715430862 - 5 \times 5.963927855711423 = -29.675351701402807 \leq 5 \quad \text{(Satisfies)} \\
4. & \quad x_1, x_2 \geq 0 \\
   & \quad 0.14428857715430862 \geq 0 \quad \text{and} \quad 5.963927855711423 \geq 0 \quad \text{(Satisfies)}
\end{align*}

# Objective Functions Evaluation

Calculate the values of the objective functions using the obtained values:

\begin{align*}
f_1 &= x_1 + 4x_2 \\
    &= 0.14428857715430862 + 4 \times 5.963927855711423 \\
    &= 24 \\
f_2 &= bx_1 + \frac{b}{2}x_2 \\
    &= 1 \times 0.14428857715430862 + \frac{1}{2} \times 5.963927855711423 \\
    &= 3.12625250501002
\end{align*}

Comparing these values with the results obtained:



*   Best 𝑓1:24.0
*   Best 𝑓2:3.12625250501002



These results confirm that the optimization process correctly maximized the objective functions within the specified constraints. The concession method applied a 7% concession, ensuring a balanced trade-off between the two objectives. Overall, the solution provided is valid and meets the criteria set out by the problem, demonstrating the effectiveness of the approach taken.